### ISSUE TITLE
Rounding issues caused by decimal limit when creating orders.

### ISSUE THREAD

**Issue by gevasi** [woocommerce/woocommerce#17637]
I heave my decimal limit set to 2 and everything works as expected, but for one of my flatfee shippingcosts I have it set to 3.264, giving my 3.95 rounded after 21% taxes.
If I just set it to 3.26 it rounds to 3.94.

When an order is created all taxes and amounts are calculated correctly giving my 3.95, but when I edit something in the order (add or delete products and such) and then recalculate the totals the shippingcost totals to 3.94. 
If I check the shippingcost in the ordermeta itself, it's only 3.26. This happens because the shippingcost is rounded to 2 decimals when the order is created. 
If I set shippingcost to 3.266 it's rounded up to 3.27, so it's not cutting of the last decimal, but rounding it.
Taxes are initially correct for the 3.264 amount used in the settings, but are updated using the 3.26 value stored in ordermeta after recalculating.
When I set the decimal limit to 3 it's stored as 3.264 in ordermeta.
I think this is a weird behaviour since all other amounts (products and taxes) are stored as floats with 4 decimals.
I work around this by editing the shipping cost to 3.264 manually when I changed anything on an order, but it's just strange behaviour since other amounts are not stored rounded. The shippingcost is not rounded again after saving, this only happens when creating the initial order.

When I verify this in the database I see 2 amounts stored rounded to the decimal limit I set in the settings: this shipping cost and the shipping_tax_amount (total tax for all shipping) 
The tax for the shippingcost line item itself is stored with 4 decimals.
After updating shippingcost to a different amount and recalculating taxes the shipping_tax_amount is also stored with 4 decimals in the database. 
Proving again that the decimal limit and storing the amounts rounded only happens upon initial order creation.

My productprices are entered with taxes included if this has any influance on this.

## Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [X] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [X] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1.Set decimal limit to 2
2.Set flat fee shipping for the appropriate shipping zone to an amount with 3 decimals (I use 3.264) (Tax is set to 21% for both products and shipping)
3.Make an order in the front-end and use a paymentmethod such as bank-transfer so the order doesn't go to the processing state.
4.Open the order in the back-end and edit the shippingcost line item. 
You should see amount 3.26 and tax amount 0.6854 (which is not the correct tax amount for 3.26 but for 3.264). 
In the database cost is also stored as 3.26 and shipping_tax_amount as 0.69.
5. If you just save and update, the tax amount is updated to 0.6846 and so is shipping_tax_amount in the database
6. If you change the shippingcost to 3.264 and recalculate everything is stored correctly again and no amounts are stored rounded in the database. They are shown rounded in the back-end but when you click edit the amounts stored in the database are shown.
7. Setting decimal limit to 3 makes these 2 database-fields rounded to 3 decimals upon order creation, while all others are still stored with 4 decimals.

## Isolating the problem

- [X] This bug happens with only WooCommerce plugin active
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [X] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
`
### WordPress Environment ###

Home URL: 
Site URL: 
WC Version: 3.2.0
Log Directory Writable: ✔
WP Version: 4.8.3
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: –
Language: nl_NL

### Server Environment ###

Server Info: Apache
PHP Version: 7.0.20
PHP Post Max Size: 256 MB
PHP Time Limit: 300
PHP Max Input Vars: 20000
cURL Version: 7.52.1
OpenSSL/1.0.1t

SUHOSIN Installed: –
MySQL Version: 5.7.18
Max Upload Size: 256 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.2.0
WC Database Prefix: hwh_
Totale databasegrootte: 18.57MB
Database datagrootte: 13.35MB
Database index grootte: 5.22MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_order_itemmeta: Data: 0.06MB + Index: 0.03MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB
wp_comments: Data: 0.02MB + Index: 0.09MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_lockdowns: Data: 0.02MB + Index: 0.00MB
wp_login_fails: Data: 0.02MB + Index: 0.00MB
wp_mollie_pending_payment: Data: 0.02MB + Index: 0.00MB
wp_options: Data: 1.52MB + Index: 0.06MB
wp_postmeta: Data: 4.52MB + Index: 3.03MB
wp_posts: Data: 5.52MB + Index: 0.39MB
wp_social_users: Data: 0.02MB + Index: 0.02MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.02MB + Index: 0.03MB
wp_term_relationships: Data: 0.08MB + Index: 0.05MB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wp_testimonial_basics: Data: 0.05MB + Index: 0.00MB
wp_usermeta: Data: 0.03MB + Index: 0.03MB
wp_users: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_bundled_itemmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_bundled_items: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_termmeta: Data: 0.02MB + Index: 0.03MB

### Post Type Counts ###

attachment: 1097
blocks: 4
mc4wp-form: 2
nav_menu_item: 52
page: 30
post: 14
product: 164
product_variation: 649
revision: 740
scheduled-action: 2
shop_coupon: 13
shop_order: 13
shop_webhook: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (2) ###

Really Simple SSL: door Rogier Lankhorst – 2.5.22
WooCommerce: door Automattic – 3.3-dev

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: EUR (€)
Currency Position: left_space
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: bundle (bundle)
composite (composite)
external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

### WC Pages ###

Winkelbasis: #2670 - /producten/
Winkelmand: #5 - /winkelmandje/
Afrekenen: #6 - /afrekenen/
Mijn account: #7 - /mijn-account/
Algemene voorwaarden: #157 - /algemene-voorwaarden/

### Theme ###

Name: Twenty Sixteen
Version: 1.3
Author URL: https://nl.wordpress.org/
Child Theme: ❌ – Als je wijzigingen aanbrengt in WooCommerce of in een hoofdthema dat je niet zelf gebouwd hebt
raden we het gebruik van een child thema aan. Zie: Hoe maak je een child thema

WooCommerce Support: ✔

### Templates ###

Overrides: –
`
</details>

**Comment by mikejolley** [comment]
Yes this is an issue. Unfortunately this cannot be fixed without introducing a way of defining shipping costs including taxes (which itself is a massive task and one with numerous backwards compatibility issues). This is not a new issue https://github.com/woocommerce/woocommerce/issues/11347

> If I check the shippingcost in the ordermeta itself, it's only 3.26. This happens because the shippingcost is rounded to 2 decimals when the order is created.

Shipping costs are indeed rounded to prevent issues when sending prices to gateways which cannot handle more than 2dp and/or need taxes to be sent separately to the cost. 

We have no immediate plans to allow shipping costs to be defined inclusive of tax. In your case you'll need to work with prices excluding tax and decide if you want shipping to charge the customer 3.94, 3.96, or manually correct edited orders.

Thanks

**Comment by gevasi** [comment]
OK, I'll keep using my workaround then.
The issue with the gateways didn't occur to me because I don't have that problem, but I see your point.
So when you're using 3 of 4 decimals globaly you have to make sure the other gateways can handle it also because the rounding to 2 decimals doesn't occur then.

It was the helpdesk that told me to post this here when I also reported the rounding bug with the coupons.
Thx for the info on why this is done the way it is.

**Comment by B1-0S** [comment]
Ha I just had a great idea, you can remove the option for "Yes, I will enter prices inclusive of tax", until fixed. That's the issue isn't or the problem is the same for the other option?

**Comment by mikejolley** [comment]
Unless you present a test case I cannot really take that report seriously. We have unit tests running on basic calculations like that, and they pass.

**Comment by B1-0S** [comment]
I will submit a valid test case. Please exuse me for my otburst earlier. I have tracked the issue and I have some serious intention to step up and help with the fix if nessasary. 

The only question I have is where to find info on when the next release is sceduled?

And you know this plugin is so awersome because of you guys!

**Comment by mikejolley** [comment]
@B1-0S Depending on the volume of fixes needing to go, out 1-2 weeks.

**Comment by B1-0S** [comment]
@mikejolley Thanks!
